### ISSUE TITLE
list(ImageSequence.Iterator(gif)) does not work as expected

### ISSUE THREAD

**Issue by readygo67** [python-pillow/Pillow#7022]
### What did you do?
split 70 frames xx.gif to several single frame .gif files.

### What did you expect to happen?
after splitted, each file is different from others. 

### What actually happened?
In the below 3 methods, all files derived from method1 are same and equal to the last frame of original .gif. method2/3 works as expected. could any one help to explain it?

### What are your OS, Python and Pillow versions?

* OS:  macOS Monterey 12.0.1
* Python:  python3 3.9 
* Pillow:  9.4.0

```python

from PIL import Image, ImageDraw, ImageFont, ImageSequence,ImageChops

def test1(text):
    with Image.open('zootopia-smile.gif') as gif:
        iters = ImageSequence.Iterator(gif)

        frames = list(iters)
        i=0
        for e in frames:
            e.save(f"frame1_{i}.gif")
            i += 1

def test2(text):
    with Image.open('zootopia-smile.gif') as gif:
        frames = ImageSequence.all_frames(gif)

        i = 0
        for e in frames:
            e.save(f"frame2_{i}.gif")
            i += 1


def test3(text):
    with Image.open('zootopia-smile.gif') as gif:
        frames = [frame.copy() for frame in ImageSequence.Iterator(gif)]
        i = 0
        for e in frames:
            e.save(f"frame3_{i}.gif")
            i += 1


# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    test1("")
    test2("")
    test3("")
```

![zootopia-smile](https://user-images.githubusercontent.com/78890754/225962436-e0374b15-eebe-4232-8c96-c02395ee5b99.gif)

**Comment by radarhere** [comment]
As background, when Pillow seeks to a different frame in an image, it is still the same object instance.
```python
>>> from PIL import Image
>>> im = Image.open("Tests/images/iss634.gif")
>>> im
<PIL.GifImagePlugin.GifImageFile image mode=P size=245x245 at 0x104D9CB80>
>>> im.seek(1)
>>> im
<PIL.GifImagePlugin.GifImageFile image mode=RGB size=245x245 at 0x104D9CB80>
```

`ImageSequence.Iterator` takes an image, and each time it moves forward, it goes to the next frame in that single image.
https://github.com/python-pillow/Pillow/blob/2a557ceb76f1760601f96df9c6e6fb653e231a71/src/PIL/ImageSequence.py#L48-L54
```python
>>> from PIL import Image, ImageSequence
>>> im = Image.open("pillow/Pillow/Tests/images/iss634.gif")
>>> it = ImageSequence.Iterator(im)
>>> for frame in it:
...   print(frame)
... 
<PIL.GifImagePlugin.GifImageFile image mode=P size=245x245 at 0x104C894C0>
<PIL.GifImagePlugin.GifImageFile image mode=RGB size=245x245 at 0x104C894C0>
...
```
It is the same image object being given back each time. So if you store `frame` in an array in the first loop, when you hit the second loop, the `frame` image will have changed. It is not a new image instance. It is the same.

So in your first method, you've asked the iterator to completely iterate through the image, leaving it at the last frame. Then you save that image repeatedly.

As for the second and third methods, your third method is really just a simplified version of our `all_frames`, so that's the same as your second method.
https://github.com/python-pillow/Pillow/blob/2a557ceb76f1760601f96df9c6e6fb653e231a71/src/PIL/ImageSequence.py#L73

By copying the image each time after the iterator has moved forward, a new image instance is created and saved in the array. Afterwards, it is the original image which is then moved forward, and the new image is left unchanged.

This behaviour has existed since PIL. The advantage of it would be smaller memory use.

So if all you are doing is saving each image, I would recommend
```python
with Image.open('zootopia-smile.gif') as gif:
    i=0
    for e in ImageSequence.Iterator(gif):
        e.save(f"frame4_{i}.gif")
        i += 1
```

**Comment by readygo67** [comment]
In my mind, apply list to an iterator will not change iterator's output
for example,
```
if __name__ == '__main__':
    list1 = [1,2,3,4,"hello"]

    my_iter = iter(list1)  # change list to iterator
    list2 = list(my_iter)  # change back to list
    print("list 2")
    for l2 in list2:
        print(l2)
```
the output is
`
list 2
1
2
3
4
hello
`
so I think apply 
`
frames = list(iters)
`
will output each frame like above case.  
In deed, I don't understand  "So in your first method, you've asked the iterator to completely iterate through the image, leaving it at the last frame. Then you save that image repeatedly.".  Could you give me more info?

**Comment by radarhere** [comment]
I'll start out with an ultra simple principle.
```pycon
>>> from PIL import Image
>>> a = Image.new("RGB", (1, 1))
>>> b = [a]
>>> a.info["test"] = "this is a test"
>>> a.info
{'test': 'this is a test'}
>>> b[0].info
{'test': 'this is a test'}
```
In that code, there is only one Image instance, `a`. Just because I put it into a list called `b`, there was not a copy created of the Image instance. When I changed `a`, the first item in the `b` list also changed. This probably seems so simple but I didn't need to say it, but on the off chance that it wasn't, there you go.

Ok, moving on.

In the example of `[1,2,3,4,"hello"]`, there are four integers and a string.

In the case of `ImageSequence.Iterator(gif)`, there is one Image instance. This not not a list of any kind. Pillow does not have all of the frames of the image loaded into memory when you start, hiding them away in some internal array to be slowly dispensed. Pillow actually reads in the next frame from the file only when the Image instance is told to seek forward. Throughout the iteration, whether you cast it to a list or not, there is only ever one Image instance.

Don't think of the iterator as working its way through a list. Imagine that you said that you wanted to deal with a bucket with one bean in it, and then a deal with a bucket with two beans in it. I could buy two buckets, but that seems expensive. Instead, I give you a bucket with one bean in it, and then when you're done, I take the bucket back and put a second bean in it and give it to you again. In this story, there is only ever one bucket. At the end (after you have finished casting to a list), if you asked me how many beans are in the bucket that I gave you at first, the answer is two - because I put a second bean in it later on. Then if you asked me how many beans are in the bucket that I gave you next, there are two. For both questions, the bucket has the same number of beans (both items in the list are at the last frame).

If we imagine an image with just two frames, then when you cast `ImageSequence.Iterator(gif)` to a list,
https://github.com/python-pillow/Pillow/blob/2a557ceb76f1760601f96df9c6e6fb653e231a71/src/PIL/ImageSequence.py#L48-L54
the iterator tells the Image instance to seek to the first frame, and then returns the Image instance to be part of the list.
Then the iterator tells the Image instance given to it to seek to the second frame. However, this is the same Image instance that is now a part of the list. So the one Image instance, the Image instance that is within the iterator and in the list, seeks to the second frame. Then that Image instance is returned to be the next part of the list.

So in the end, the list has two references to the same Image instance, which is seeked to the second frame.

From that, calling `list(ImageSequence.Iterator(gif))` is not going to work, because it is returning a list of multiple references to the same Image instance.

This behaviour has existed since Pillow was forked from PIL. The advantage of the current method would be smaller memory use, because something like
```python
for e in ImageSequence.Iterator(gif):
    e.save(f"frame4_{i}.gif")
```
doesn't need multiple Image instances. It simply performs an operation on the Image instance in the current state and moves on.
